charts/redpanda/templates/_helpers.tpl

```diff
@@ -869,25 +869,79 @@ REDPANDA_SASL_USERNAME REDPANDA_SASL_PASSWORD REDPANDA_SASL_MECHANISM
   {{- end -}}
 {{- end -}}
 
-{{- define "storage-tiered-credentials-secret-key" -}}
-{{- $oldCondtion := (and .Values.storage.tiered.credentialsSecretRef.name .Values.storage.tiered.credentialsSecretRef.key) -}}
-{{- $newCondtion := (and .Values.storage.tiered.credentialsSecretRef.secretKey.name .Values.storage.tiered.credentialsSecretRef.secretKey.key) -}}
-{{- $configurationKey := (dig "configurationKey" "" .Values.storage.tiered.credentialsSecretRef) -}}
-{{- if empty $configurationKey -}}
-  {{- $configurationKey = .Values.storage.tiered.credentialsSecretRef.secretKey.configurationKey -}}
-{{- end -}}
-{{- $key := (dig "key" "" .Values.storage.tiered.credentialsSecretRef) -}}
-{{- if empty $key -}}
-  {{- $key = .Values.storage.tiered.credentialsSecretRef.secretKey.key -}}
-{{- end -}}
-{{- $name := (dig "name" "" .Values.storage.tiered.credentialsSecretRef) -}}
-{{- if empty $name -}}
-  {{- $name = .Values.storage.tiered.credentialsSecretRef.secretKey.name -}}
-{{- end -}}
-{{- toJson (dict
-  "bool" (or $oldCondtion $newCondtion)
-  "configurationKey" $configurationKey
-  "key" $key
-  "name" $name
-) -}}
+{{/* secret-ref-or-value
+        in: {Value: string?, SecretKey: string?, SecretName: string?}
+        out: corev1.Envvar | nil
+    secret-ref-or-value converts a set of values into a structure suitable for
+    use as an environment variable or nil.
+*/}}
+{{- define "secret-ref-or-value" -}}
+    {{- if and (empty .Value) (or (empty .SecretName) (empty .SecretKey)) -}}
+        {{- mustToJson nil -}}
+    {{- else -}}
+        {{- $out := (dict
+            "name" .Name
+            "value" .Value
+            "valueFrom" (dict
+                "secretKeyRef" (dict
+                    "name" .SecretName
+                    "key" .SecretKey
+                )
+            )
+        ) -}}
+        {{- if empty .Value -}}
+            {{- $_ := unset $out "value" -}}
+        {{- else -}}
+            {{- $_ := unset $out "valueFrom" -}}
+        {{- end -}}
+        {{- mustToJson $out -}}
+    {{- end -}}
+{{- end -}}
+
+{{- define "tiered-storage-env-vars" -}}
+    {{- $config := (include "storage-tiered-config" . | fromJson) -}}
+    [
```

Comment:
Maybe it would be easier to create list of corev1.Envvar using `prepend` or `append` and then pipe it to `toJson` to not deal with commas.